unit-tests

hifeful opened the pull request:
- unit tests for a `MealManiaViewModel`, `MealDetailsViewState`, and `Reducer` of the `MealDetailsFeature` were added

hifeful committed:
- unit tests for a MealManiaViewModel, MealDetailsViewState, and Reducer of the MealDetailsFeature were added

epam-alyse commented on app/src/test/java/com/hifeful/mealmania/presentation/MealManiaViewModelTest.kt:
```diff
@@ -0,0 +1,121 @@
+    private val viewModel = MealManiaViewModel()
+
+    @Test
+    fun `Given mymeals page deeplink, When handleDeeplink is called, Then ShowMyMealsPage DeeplinkResult should be returned`() {
```
I would suggest using the following template:
`with mymeals page deeplink - on handleDeeplink - it should return DeeplinkResult.ShowMyMealsPage`

hifeful replied:
Good suggestion. But I used that naming from my current project. We're trying to use **TDD's** approach to test naming.

Your approach is similar to mine. **With** instead of **Given**, **On** instead of **When**, and **Then** is just omitted.

epam-alyse commented on app/src/test/java/com/hifeful/mealmania/presentation/MealManiaViewModelTest.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Test
+    fun `Given not supported deeplink, When handleDeeplink is called, Then null should be returned`() {
```
`with unsupported deeplink - on handleDeeplink - it should return no deeplink`

hifeful replied:
Changed.

epam-alyse commented on app/src/test/java/com/hifeful/mealmania/presentation/MealManiaViewModelTest.kt:
```diff
@@ -0,0 +1,121 @@
+        val actual = viewModel.handleDeeplink(uri)
+
+        // Then
+        assert(actual is DeeplinkResult.ShowMyMealsPage)
```
I know the handy API is provided by the library `org.assertj.core.api.Assertions`:
isEqualTo, isNull, isEmpty, etc.

hifeful replied:
Thank you for the suggestion! 
Usually, I'm using **Google's Truth** library for **assertion** purposes. But I checked and used your library and it's handy to use.

epam-alyse commented on app/src/test/java/com/hifeful/mealmania/presentation/MealManiaViewModelTest.kt:
```diff
@@ -0,0 +1,121 @@
+        val actual = viewModel.handleDeeplink(uri)
+
+        // Then
+        assert(actual == null)
```
So, comments can be removed

hifeful replied:
Of course, they can be removed. It's one more approach that is used on my current project. We're adding comments to _logically separate_ code related to the TDD's **Given**, **When**, **Then** sections.

epam-alyse replied:
In my experience, we usually use an empty next line to separate blocks 🙂

epam-alyse commented on app/src/test/java/com/hifeful/mealmania/presentation/details/MealDetailsViewStateTest.kt:
```diff
@@ -0,0 +1,76 @@
+            drinkAlternate = drinkAlternate,
+            youtubeSource = youtubeSource,
+            ingredients = ingredients
+        )
```
Usually, it's avoided and mocks are used instead. For example, Mockito library can be used where we can provide relaxed objects: mock(RETURN_MOCKS). There 4 different flags for the mock configuration

hifeful replied:
Good suggestion. But I prefer to use **mockK** library instead of **Mockito**. 👍

epam-alyse reviewed (commented):
WDYT are the task requirements exactly the way of TDD?

epam-alyse reviewed (changes requested):
Needs to demonstrate usage of Robolectric

hifeful committed:
- AssertJ and mockK libraries were added and used for the unit tests
- mocks instead of real objects are used in the MealDetailsViewStateTest now
- test titles in the MealManiaViewModelTest.kt were changed
- comments were removed

hifeful commented:
> WDYT are the task requirements exactly the way of TDD?

If we're talking about the strict **TDD**, then we should start writing our code from the **unit tests**. 
So, we should write tests that don't compile regarding our business logic and then implement the actual code.
Unfortunately, I didn't work on projects like that and I'm not quite sure if it's really needed.

hifeful commented:
> Needs to demonstrate usage of Robolectric

Regarding the usage of **Robolectic**, in my **unit tests**, I added `@RunWith` annotation with the **AndroidJUnit4** runner that makes available classes from the **Android SDK**. I needed `Uri` and `Context` classes, so **Robolectric** helped me.
